programs/ssl/ssl_server2.c

```diff
@@ -3742,6 +3745,12 @@ int main( int argc, char *argv[] )
     }
 #endif /* MBEDTLS_SSL_DTLS_CONNECTION_ID */
 
+#if defined(MBEDTLS_MEMORY_DEBUG)
+    mbedtls_memory_buffer_alloc_cur_get(&current_heap_memory, &heap_blocks);
+    mbedtls_memory_buffer_alloc_max_get(&peak_heap_memory, &heap_blocks);
+    mbedtls_printf("Heap memory usage after handshake: %u bytes. Peak memory usage was %u\n", current_heap_memory, peak_heap_memory);
```

Comment:
I get warnings here when building on Linux amd64. Even though casts are ugly, I think we might need to cast the `size_t` to an `unsigned long` an use `%lu` to avoid warnings on all platforms.

```
programs/ssl/ssl_server2.c: In function ‘main’:
programs/ssl/ssl_server2.c:3751:57: warning: format ‘%u’ expects argument of type ‘unsigned int’, but argument 2 has type ‘size_t {aka long unsigned int}’ [-Wformat=]
     mbedtls_printf("Heap memory usage after handshake: %u bytes. Peak memory usage was %u\n", current_heap_memory, peak_heap_memory);
                                                        ~^
                                                        %lu
programs/ssl/ssl_server2.c:3751:89: warning: format ‘%u’ expects argument of type ‘unsigned int’, but argument 3 has type ‘size_t {aka long unsigned int}’ [-Wformat=]
     mbedtls_printf("Heap memory usage after handshake: %u bytes. Peak memory usage was %u\n", current_heap_memory, peak_heap_memory);
                                                                                        ~^
                                                                                        %lu
```

Reply:
If we required C99 support, we could use `%zu`.

By the way, what is the status of C89/C99 support in Mbed TLS? IIRC then we started using C99 in Mbed Crypto already, but we still support C89 in Mbed TLS. Am I remembering this correctly?

Reply:
We require a subset of C99 for both Mbed Crypto and Mbed TLS. However, C library support for compilers varies, even if they do support the C99 language. Not all `printf()` will have `%zu` available on embedded systems. So, we are still conservative about what we use in order to maximize portability. This is up to date (Sept 2018) https://tls.mbed.org/kb/development/mbedtls-coding-standards

Reply:
hmm... I did not get this warning on my "gcc version 7.4.0 (Ubuntu 7.4.0-1ubuntu1~18.04.1)", but I have changed it to %lu